src/Temporalio/Client/Configuration/DataSource.cs

```diff
@@ -0,0 +1,151 @@
+using System;
+using System.Text;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Temporalio.Client.Configuration
+{
+    /// <summary>
+    /// A data source for configuration, which can be a path to a file,
+    /// the string contents of a file, or raw bytes.
+    /// </summary>
+    [JsonConverter(typeof(DataSource.JsonConverter))]
+    public sealed class DataSource
+    {
+        private DataSource()
+        {
+        }
+
+        /// <summary>
+        /// Gets the file path for this data source, if applicable.
+        /// </summary>
+        public string? Path { get; private set; }
+
+        /// <summary>
+        /// Gets the raw data for this data source, if applicable.
+        /// </summary>
+        public byte[]? Data { get; private set; }
```

Comment:
```suggestion
        public string? Path { get; private init; }

        /// <summary>
        /// Gets the raw data for this data source, if applicable.
        /// </summary>
        public byte[]? Data { get; private init; }
```

These should only be mutable on init